File: backend/src/modloader/forge.ts
import got, { Got } from "got";
import fs from "node:fs"
import fse from "fs-extra"
import { execPromise, fastdownload, version_compare, verifySHA1 } from "../utils/utils.js";
import { Azip } from "../utils/ziplib.js";
import { execSync } from "node:child_process";
import config from "../utils/config.js";
import { logger } from "../utils/logger.js";

interface Iforge{
    data:{
        MOJMAPS:{
            server:string
        },
        MAPPINGS:{
            server:string
        }
    }
}

interface Iversion{
    downloads:{
        server_mappings:{
            url:string
        }
    }
}

interface IForgeFile {
    format: string;
    category: string;
    hash: string;
    _id: string;
}

interface IForgeBuild {
    branch: string;
    build: number;
    mcversion: string;
    modified: string;
    version: string;
    _id: string;
    files: IForgeFile[];
}

export class Forge {
    minecraft: string;
    loaderVersion: string;
    got: Got;
    path: string;
 constructor(minecraft:string,loaderVersion:string,path:string){
    this.minecraft = minecraft;
    this.loaderVersion = loaderVersion;
    this.path = path
    this.got = got.extend({
        headers: { "User-Agent": "DeEarthX" },
        hooks: {
            init: [
                (options) => {
                    if(config.mirror.bmclapi){
                        options.prefixUrl = "https://bmclapi2.bangbang93.com";
                    }else{
                        options.prefixUrl = "http://maven.minecraftforge.net/";
                    }
                }
            ]
        }
    })
 }

 async setup(){
    await this.installer()
    if(config.mirror.bmclapi){
        await this.library()
    }
    await this.install()
    if (version_compare(this.minecraft,"1.18") === -1){
        await this.wshell()
    }
 }

 async library(){
    const _downlist: [string,string][]= []
    const data = await fs.promises.readFile(`${this.path}/forge-${this.minecraft}-${this.loaderVersion}-installer.jar`)
    const zip = Azip(data)
    for await(const entry of zip){
        if(entry.entryName === "version.json" || entry.entryName === "install_profile.json"){ //Libraries
        JSON.parse((entry.getData()).toString()).libraries.forEach(async (e:any)=>{
            const t = e.downloads.artifact.path
            _downlist.push([`https://bmclapi2.bangbang93.com/maven/${t}`,`${this.path}/libraries/${t}`])
        })
        }
        if(entry.entryName === "install_profile.json"){ //MOJMAPS
            /* 获取MOJMAPS */
            const json = JSON.parse((entry.getData()).toString()) as Iforge
            const vjson = await this.got.get(`version/${this.minecraft}/json`).json<Iversion>()
            console.log(`${new URL(vjson.downloads.server_mappings.url).pathname}`)
            const mojpath = this.MTP(json.data.MOJMAPS.server)
            _downlist.push([`https://bmclapi2.bangbang93.com/${new URL(vjson.downloads.server_mappings.url).pathname.slice(1)}`,`${this.path}/libraries/${mojpath}`])
            /* 获取MOJMAPS */

            /*获取MAPPING*/
            const mappingobj = json.data.MAPPINGS.server
            const path = this.MTP(mappingobj.replace(":mappings@txt","@zip"))
            _downlist.push([`https://bmclapi2.bangbang93.com/maven/${path}`,`${this.path}/libraries/${path}`])
            /* 获取MAPPING */
        }
    }
    const downlist = [...new Set(_downlist)]
    await fastdownload(downlist)
 }

 async install(){
   const javaCmd = config.javaPath || 'java';
   await execPromise(`${javaCmd} -jar forge-${this.minecraft}-${this.loaderVersion}-installer.jar --installServer`,{cwd:this.path})
 }

async installer(){
    let url = `forge/download?mcversion=${this.minecraft}&version=${this.loaderVersion}&category=installer&format=jar`
    let expectedHash: string | undefined;
    
    // 如果使用 BMCLAPI，先获取版本信息以获取 hash
    if (config.mirror?.bmclapi) {
        try {
            const forgeInfo = await this.got.get(`forge/minecraft/${this.minecraft}`).json<IForgeBuild[]>();
            // 查找匹配的 forge 版本
            const forgeVersion = forgeInfo.find(f => f.version === this.loaderVersion);
            if (forgeVersion) {
                // 查找 installer 文件的 hash
                const installerFile = forgeVersion.files.find(f => f.category === 'installer' && f.format === 'jar');
                if (installerFile) {
                    expectedHash = installerFile.hash;
                    logger.debug(`获取到 Forge installer hash: ${expectedHash}`);
                }
            }
        } catch (error) {
            logger.warn(`获取 Forge hash 信息失败，将跳过 hash 验证`, error);
        }
    } else {
        url = `net/minecraftforge/forge/${this.minecraft}-${this.loaderVersion}/forge-${this.minecraft}-${this.loaderVersion}-installer.jar`
    }
    
    const res = (await this.got.get(url)).rawBody;
    const filePath = `${this.path}/forge-${this.minecraft}-${this.loaderVersion}-installer.jar`;
    await fse.outputFile(filePath, res);
    
    // 如果获取到了 hash，验证下载的文件
    if (expectedHash) {
        if (!verifySHA1(filePath, expectedHash)) {
            // hash 验证失败，删除文件并重新下载
            logger.warn(`Forge installer hash 验证失败，删除文件并重试`);
            fs.unlinkSync(filePath);
            // 重新下载一次
            const res2 = (await this.got.get(url)).rawBody;
            await fse.outputFile(filePath, res2);
            
            // 再次验证
            if (!verifySHA1(filePath, expectedHash)) {
                throw new Error(`Forge installer hash 验证失败，文件可能已损坏`);
            }
        }
    }
}

 private async wshell(){
    const javaCmd = config.javaPath || 'java';
    const cmd = `${javaCmd} -jar forge-${this.minecraft}-${this.loaderVersion}.jar`
    await fs.promises.writeFile(`${this.path}/run.bat`,`@echo off\n${cmd}`) //Windows
    await fs.promises.writeFile(`${this.path}/run.sh`,`#!/bin/bash\n${cmd}`) //Linux
 }

 private MTP(string:string){
    const mjp =  string.replace(/^\[|\]$/g, '')
    const OriginalName = mjp.split("@")[0]
    const x = OriginalName.split(":")
    const mappingType = mjp.split('@')[1];
    if(x[3]){
    return `${x[0].replace(/\./g, '/')}/${x[1]}/${x[2]}/${x[1]}-${x[2]}-${x[3]}.${mappingType}`
    }else{
    return `${x[0].replace(/\./g, '/')}/${x[1]}/${x[2]}/${x[1]}-${x[2]}.${mappingType}`
    }
  }    
}

File: backend/modloader/forge.go
package modloader

import (
	"log/slog"
	"archive/zip"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"dex/backend/utils"

	"resty.dev/v3"
)

func NewForge(mcv, mlv, path string) *Forge {
	f := &Forge{
		minecraft:     mcv,
		loaderVersion: mlv,
		path:          path,
	}

	bmclapi := utils.GlobalConfig.GetConfigValue("mirror.bmclapi")
	baseURL := "http://maven.minecraftforge.net/"
	if bmclapi == true {
		baseURL = "https://bmclapi2.bangbang93.com/"
	}

	f.client = resty.New().
		SetBaseURL(baseURL).
		SetHeader("User-Agent", "DeEarthX")

	return f
}

type Forge struct {
	minecraft     string
	loaderVersion string
	path          string
	client        *resty.Client
}

func (f *Forge) Setup() error {
	// Step 1: Download installer
	if err := f.Installer(); err != nil {
		return fmt.Errorf("forge installer download failed: %w", err)
	}

	// Step 2: Download libraries (if BMCLAPI mirror enabled and MC > 1.10)
	bmclapi := utils.GlobalConfig.GetConfigValue("mirror.bmclapi")
	if bmclapi == true && utils.VersionCompare(f.minecraft, "1.10") == 1 {
		if err := f.Library(); err != nil {
			return fmt.Errorf("forge library download failed: %w", err)
		}
	}

	// Step 3: Run installer
	if err := f.Install(); err != nil {
		return fmt.Errorf("forge install failed: %w", err)
	}

	// Step 4: Write shell scripts for MC < 1.18
	if utils.VersionCompare(f.minecraft, "1.18.0") == -1 {
		if err := f.wshell(); err != nil {
			return fmt.Errorf("forge shell script generation failed: %w", err)
		}
	}

	return nil
}

func (f *Forge) Installer() error {
	bmclapi := utils.GlobalConfig.GetConfigValue("mirror.bmclapi")

	relURL := fmt.Sprintf("forge/download?mcversion=%s&version=%s&category=installer&format=jar", f.minecraft, f.loaderVersion)
	if utils.VersionCompare(f.minecraft, "1.10") != 1 {
		relURL += "&branch=" + f.minecraft
	}

	var expectedHash string

	if bmclapi == true {
		// Try to get hash from BMCLAPI forge info
		forgeInfoURL := fmt.Sprintf("forge/minecraft/%s", f.minecraft)
		resp, err := f.client.R().Get(forgeInfoURL)
		if err == nil && resp.StatusCode() < 400 {
			var forgeBuilds []ForgeBuild
			if err := parseJSON(resp.Bytes(), &forgeBuilds); err == nil {
				for _, build := range forgeBuilds {
					if build.Version == f.loaderVersion {
						for _, file := range build.Files {
							if file.Category == "installer" && file.Format == "jar" {
								expectedHash = file.Hash
								break
							}
						}
					}
				}
			}
		}
	} else {
		// Official Maven URL
		relURL = fmt.Sprintf("net/minecraftforge/forge/%s-%s/forge-%s-%s-installer.jar",
			f.minecraft, f.loaderVersion, f.minecraft, f.loaderVersion)
	}

	// Build full URL for chunked download
	baseURL := f.client.BaseURL()
	fullURL := baseURL + relURL

	filePath := filepath.Join(f.path, fmt.Sprintf("forge-%s-%s-installer.jar", f.minecraft, f.loaderVersion))
	if err := os.MkdirAll(filepath.Dir(filePath), 0o755); err != nil {
		return err
	}

	// Use chunked download for the installer jar
	downloadClient := utils.NewDownloadClient()
	if expectedHash != "" {
		if err := downloadClient.ChunkedDownload(fullURL, filePath, expectedHash); err != nil {
			// Hash verification failed, retry once
			slog.Warn("Forge installer hash verification failed, deleting and retrying")
			os.Remove(filePath)

			if err := downloadClient.ChunkedDownload(fullURL, filePath, expectedHash); err != nil {
				return fmt.Errorf("forge installer hash verification failed after retry: %w", err)
			}
		}
	} else {
		if err := downloadClient.ChunkedDownload(fullURL, filePath); err != nil {
			return fmt.Errorf("failed to download forge installer: %w", err)
		}
	}

	return nil
}

func (f *Forge) Library() error {
	var downloadItems []utils.DownloadOption

	installerPath := filepath.Join(f.path, fmt.Sprintf("forge-%s-%s-installer.jar", f.minecraft, f.loaderVersion))
	r, err := zip.OpenReader(installerPath)
	if err != nil {
		return fmt.Errorf("failed to open forge installer jar: %w", err)
	}
	defer r.Close()

	for _, file := range r.File {
		if file.Name == "version.json" || file.Name == "install_profile.json" {
			data, err := readZipFile(file)
			if err != nil {
				continue
			}

			var profile struct {
				Libraries []struct {
					Downloads struct {
						Artifact struct {
							Path string `json:"path"`
						} `json:"artifact"`
					} `json:"downloads"`
				} `json:"libraries"`
				Data struct {
					MOJMAPS struct {
						Server string `json:"server"`
					} `json:"MOJMAPS"`
					MAPPINGS struct {
						Server string `json:"server"`
					} `json:"MAPPINGS"`
				} `json:"data"`
			}

			if err := parseJSON(data, &profile); err != nil {
				continue
			}

			// Add library downloads
			for _, lib := range profile.Libraries {
				if lib.Downloads.Artifact.Path != "" {
					libPath := lib.Downloads.Artifact.Path
					downloadItems = append(downloadItems, utils.DownloadOption{
						URL:      "https://bmclapi2.bangbang93.com/maven/" + libPath,
						FilePath: filepath.Join(f.path, "libraries", libPath),
					})
				}
			}

			// Handle install_profile.json specific logic
			if file.Name == "install_profile.json" {
				// NeoForge 26.x+ doesn't need mappings
				if utils.VersionCompare(f.loaderVersion, "26") >= 0 {
					continue
				}

				// MC >= 1.18: download MOJMAPS
				if utils.VersionCompare(f.minecraft, "1.18") >= 0 && profile.Data.MOJMAPS.Server != "" {
					mojPath := MTP(profile.Data.MOJMAPS.Server)

					// Fetch version JSON to get server_mappings URL
					resp, err := f.client.R().Get("version/" + f.minecraft + "/json")
					if err == nil && resp.StatusCode() < 400 {
						var versionJSON struct {
							Downloads struct {
								ServerMappings struct {
									URL string `json:"url"`
								} `json:"server_mappings"`
							} `json:"downloads"`
						}
						if err := parseJSON(resp.Bytes(), &versionJSON); err == nil {
							parsedURL, _ := url.Parse(versionJSON.Downloads.ServerMappings.URL)
							if parsedURL != nil {
								downloadItems = append(downloadItems, utils.DownloadOption{
									URL:      "https://bmclapi2.bangbang93.com/" + strings.TrimPrefix(parsedURL.Path, "/"),
									FilePath: filepath.Join(f.path, "libraries", mojPath),
								})
							}
						}
					}
				}

				// MC > 1.12.2: download MCP Mappings
				if utils.VersionCompare(f.minecraft, "1.12.2") == 1 && profile.Data.MAPPINGS.Server != "" {
					mappingPath := MTP(strings.Replace(profile.Data.MAPPINGS.Server, ":mappings@txt", "@zip", 1))
					downloadItems = append(downloadItems, utils.DownloadOption{
						URL:      "https://bmclapi2.bangbang93.com/maven/" + mappingPath,
						FilePath: filepath.Join(f.path, "libraries", mappingPath),
					})
				}
			}
		}
	}

	// Deduplicate download items
	downloadItems = dedupDownloadItems(downloadItems)

	return utils.FastDownload(downloadItems)
}

func (f *Forge) Install() error {
	javaCmd := getJavaCmd()
	cmd := fmt.Sprintf("%s -jar forge-%s-%s-installer.jar --installServer", javaCmd, f.minecraft, f.loaderVersion)

	bmclapi := utils.GlobalConfig.GetConfigValue("mirror.bmclapi")
	if bmclapi == true && utils.VersionCompare(f.minecraft, "1.10") == 1 {
		cmd += " --mirror https://bmclapi2.bangbang93.com/maven/"
	}

	if err := utils.ExecPromise(cmd, f.path); err != nil {
		return fmt.Errorf("forge install failed: %w", err)
	}

	return nil
}

func (f *Forge) wshell() error {
	javaCmd := getJavaCmd()
	cmd := fmt.Sprintf("%s -jar forge-%s-%s.jar", javaCmd, f.minecraft, f.loaderVersion)
	if utils.VersionCompare(f.minecraft, "1.10.0") == -1 {
		cmd = fmt.Sprintf("%s -jar forge-%s-%s-%s-universal.jar", javaCmd, f.minecraft, f.loaderVersion, f.minecraft)
	}

	batContent := fmt.Sprintf("@echo off\n%s", cmd)
	shContent := fmt.Sprintf("#!/bin/bash\n%s", cmd)

	if err := os.WriteFile(filepath.Join(f.path, "run.bat"), []byte(batContent), 0o644); err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(f.path, "run.sh"), []byte(shContent), 0o644)
}

// ForgeBuild represents a Forge build from BMCLAPI.
type ForgeBuild struct {
	Version string      `json:"version"`
	Files   []ForgeFile `json:"files"`
}

// ForgeFile represents a file within a Forge build.
type ForgeFile struct {
	Format   string `json:"format"`
	Category string `json:"category"`
	Hash     string `json:"hash"`
}

// dedupDownloadItems removes duplicate download items by URL.
func dedupDownloadItems(items []utils.DownloadOption) []utils.DownloadOption {
	seen := make(map[string]bool)
	var result []utils.DownloadOption
	for _, item := range items {
		if !seen[item.URL] {
			seen[item.URL] = true
			result = append(result, item)
		}
	}
	return result
}
